### ISSUE TITLE
swarm join container send tcp reset

### ISSUE THREAD

**Issue by maguec** [docker/swarm#915]
=== Docker running Swarm manage ===
```
docker --tlsverify --tlscacert=tcd-ca.pem --tlscert=tcd-client-cert.pem --tlskey=tcd-client-key.pem -H=swarm-tcd.sv2.example.com:2376 version
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2
OS/Arch (client): linux/amd64
Server version: swarm/0.2.0
Go version (server): go1.3.3
Git commit (server): 48fd993
OS/Arch (server): linux/amd64
docker logs:
time="2015-06-08T20:50:03Z" level=error msg="Get https://172.19.81.223:2375/v1.15/info: dial tcp 172.19.81.223:2375: connection refused"
time="2015-06-08T20:50:03Z" level=error msg="Get https://172.19.85.223:2375/v1.15/info: dial tcp 172.19.85.223:2375: connection refused"
time="2015-06-08T20:50:05Z" level=error msg="Get https://172.19.81.223:2375/v1.15/info: dial tcp 172.19.81.223:2375: connection refused"
time="2015-06-08T20:50:05Z" level=error msg="Get https://172.19.85.223:2375/v1.15/info: dial tcp 172.19.85.223:2375: connection refused"
docker inspect
[{
"AppArmorProfile": "",
"Args": [
"manage",
"--tlsverify",
"--tlscacert=/etc/docker-ssl/tcd-ca.pem",
"--tlscert=/etc/docker-ssl/tcd-server-cert.pem",
"--tlskey=/etc/docker-ssl/tcd-server-key.pem",
"-H",
"tcp://0.0.0.0:2376",
"consul://dckrm01.sv2.example.com:8500/swarm"
],
"Config": {
"AttachStderr": false,
"AttachStdin": false,
"AttachStdout": false,
"Cmd": [
"manage",
"--tlsverify",
"--tlscacert=/etc/docker-ssl/tcd-ca.pem",
"--tlscert=/etc/docker-ssl/tcd-server-cert.pem",
"--tlskey=/etc/docker-ssl/tcd-server-key.pem",
"-H",
"tcp://0.0.0.0:2376",
"consul://dckrm01.sv2.example.com:8500/swarm"
],
"CpuShares": 0,
"Cpuset": "",
"Domainname": "",
"Entrypoint": [
"/swarm"
],
"Env": [
"SWARM_HOST=:2375"
],
"ExposedPorts": {
"2375/tcp": {},
"2376/tcp": {}
},
"Hostname": "f6b1f35d7e21",
"Image": "swarm",
"Labels": {},
"MacAddress": "",
"Memory": 0,
"MemorySwap": 0,
"NetworkDisabled": false,
"OnBuild": null,
"OpenStdin": false,
"PortSpecs": null,
"StdinOnce": false,
"Tty": false,
"User": "",
"Volumes": {
"/.swarm": {}
},
"WorkingDir": ""
},
"Created": "2015-06-08T20:47:29.984633072Z",
"Driver": "aufs",
"ExecDriver": "native-0.2",
"ExecIDs": null,
"HostConfig": {
"Binds": [
"/etc/docker-ssl:/etc/docker-ssl"
],
"CapAdd": null,
"CapDrop": null,
"CgroupParent": "",
"ContainerIDFile": "/var/run/docker-swarm.cid",
"CpuShares": 0,
"CpusetCpus": "",
"Devices": [],
"Dns": null,
"DnsSearch": null,
"ExtraHosts": null,
"IpcMode": "",
"Links": null,
"LogConfig": {
"Config": null,
"Type": "json-file"
},
"LxcConf": [],
"Memory": 0,
"MemorySwap": 0,
"NetworkMode": "bridge",
"PidMode": "",
"PortBindings": {
"2376/tcp": [
{
"HostIp": "",
"HostPort": "2376"
}
]
},
"Privileged": false,
"PublishAllPorts": false,
"ReadonlyRootfs": false,
"RestartPolicy": {
"MaximumRetryCount": 0,
"Name": "no"
},
"SecurityOpt": null,
"Ulimits": null,
"VolumesFrom": null
},
"HostnamePath": "/var/lib/docker/containers/f6b1f35d7e2140238f9dceae397635ed5bf5e3c47f28e8d7595b78b235729e69/hostname",
"HostsPath": "/var/lib/docker/containers/f6b1f35d7e2140238f9dceae397635ed5bf5e3c47f28e8d7595b78b235729e69/hosts",
"Id": "f6b1f35d7e2140238f9dceae397635ed5bf5e3c47f28e8d7595b78b235729e69",
"Image": "73504b2882a32542f56bffefc4f7b7a0606b285bc819bc161356b91f0e32b4b0",
"LogPath": "/var/lib/docker/containers/f6b1f35d7e2140238f9dceae397635ed5bf5e3c47f28e8d7595b78b235729e69/f6b1f35d7e2140238f9dceae397635ed5bf5e3c47f28e8d7595b78b235729e69-json.log",
"MountLabel": "",
"Name": "/fervent_payne",
"NetworkSettings": {
"Bridge": "docker0",
"Gateway": "172.17.42.1",
"GlobalIPv6Address": "",
"GlobalIPv6PrefixLen": 0,
"IPAddress": "172.17.0.1",
"IPPrefixLen": 16,
"IPv6Gateway": "",
"LinkLocalIPv6Address": "fe80::42:acff:fe11:1",
"LinkLocalIPv6PrefixLen": 64,
"MacAddress": "02:42:ac:11:00:01",
"PortMapping": null,
"Ports": {
"2375/tcp": null,
"2376/tcp": [
{
"HostIp": "0.0.0.0",
"HostPort": "2376"
}
]
}
},
"Path": "/swarm",
"ProcessLabel": "",
"ResolvConfPath": "/var/lib/docker/containers/f6b1f35d7e2140238f9dceae397635ed5bf5e3c47f28e8d7595b78b235729e69/resolv.conf",
"RestartCount": 0,
"State": {
"Dead": false,
"Error": "",
"ExitCode": 0,
"FinishedAt": "0001-01-01T00:00:00Z",
"OOMKilled": false,
"Paused": false,
"Pid": 21117,
"Restarting": false,
"Running": true,
"StartedAt": "2015-06-08T20:47:30.118116754Z"
},
"Volumes": {
"/.swarm": "/var/lib/docker/vfs/dir/9320ffec7f3db60c38877e446d89e3837b8f11fa8213bd42b79719f0577eebf4",
"/etc/docker-ssl": "/etc/docker-ssl"
},
"VolumesRW": {
"/.swarm": true,
"/etc/docker-ssl": true
}
}
]
```

=== Docker running Swarm join ===
```
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2
OS/Arch (client): linux/amd64
Server version: 1.6.2
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 7c8fca2
OS/Arch (server): linux/amd64
$ sudo docker inspect 9c7f3e46138b
[{
"AppArmorProfile": "",
"Args": [
"join",
"--addr=172.19.81.223:2375",
"consul://dckrm01.sv2.example.com:8500/swarm"
],
"Config": {
"AttachStderr": false,
"AttachStdin": false,
"AttachStdout": false,
"Cmd": [
"join",
"--addr=172.19.81.223:2375",
"consul://dckrm01.sv2.example.com:8500/swarm"
],
"CpuShares": 0,
"Cpuset": "",
"Domainname": "",
"Entrypoint": [
"/swarm"
],
"Env": [
"SWARM_HOST=:2375"
],
"ExposedPorts": {
"2375/tcp": {}
},
"Hostname": "9c7f3e46138b",
"Image": "swarm",
"Labels": {},
"MacAddress": "",
"Memory": 0,
"MemorySwap": 0,
"NetworkDisabled": false,
"OnBuild": null,
"OpenStdin": false,
"PortSpecs": null,
"StdinOnce": false,
"Tty": false,
"User": "",
"Volumes": {
"/.swarm": {}
},
"WorkingDir": ""
},
"Created": "2015-06-08T20:47:33.340993496Z",
"Driver": "aufs",
"ExecDriver": "native-0.2",
"ExecIDs": null,
"HostConfig": {
"Binds": [
"/etc/docker-ssl:/etc/docker-ssl"
],
"CapAdd": null,
"CapDrop": null,
"CgroupParent": "",
"ContainerIDFile": "/var/run/docker-swarm.cid",
"CpuShares": 0,
"CpusetCpus": "",
"Devices": [],
"Dns": null,
"DnsSearch": null,
"ExtraHosts": null,
"IpcMode": "",
"Links": null,
"LogConfig": {
"Config": null,
"Type": "json-file"
},
"LxcConf": [],
"Memory": 0,
"MemorySwap": 0,
"NetworkMode": "bridge",
"PidMode": "",
"PortBindings": {
"2375/tcp": [
{
"HostIp": "",
"HostPort": "2375"
}
]
},
"Privileged": false,
"PublishAllPorts": false,
"ReadonlyRootfs": false,
"RestartPolicy": {
"MaximumRetryCount": 0,
"Name": "no"
},
"SecurityOpt": null,
"Ulimits": null,
"VolumesFrom": null
},
"HostnamePath": "/var/lib/docker/containers/9c7f3e46138b8ea2faa59212aaee4189165374a431797503b985cccfc075021d/hostname",
"HostsPath": "/var/lib/docker/containers/9c7f3e46138b8ea2faa59212aaee4189165374a431797503b985cccfc075021d/hosts",
"Id": "9c7f3e46138b8ea2faa59212aaee4189165374a431797503b985cccfc075021d",
"Image": "73504b2882a32542f56bffefc4f7b7a0606b285bc819bc161356b91f0e32b4b0",
"LogPath": "/var/lib/docker/containers/9c7f3e46138b8ea2faa59212aaee4189165374a431797503b985cccfc075021d/9c7f3e46138b8ea2faa59212aaee4189165374a431797503b985cccfc075021d-json.log",
"MountLabel": "",
"Name": "/stoic_poincare",
"NetworkSettings": {
"Bridge": "docker0",
"Gateway": "172.17.42.1",
"GlobalIPv6Address": "",
"GlobalIPv6PrefixLen": 0,
"IPAddress": "172.17.0.1",
"IPPrefixLen": 16,
"IPv6Gateway": "",
"LinkLocalIPv6Address": "fe80::42:acff:fe11:1",
"LinkLocalIPv6PrefixLen": 64,
"MacAddress": "02:42:ac:11:00:01",
"PortMapping": null,
"Ports": {
"2375/tcp": [
{
"HostIp": "0.0.0.0",
"HostPort": "2375"
}
]
}
},
"Path": "/swarm",
"ProcessLabel": "",
"ResolvConfPath": "/var/lib/docker/containers/9c7f3e46138b8ea2faa59212aaee4189165374a431797503b985cccfc075021d/resolv.conf",
"RestartCount": 0,
"State": {
"Dead": false,
"Error": "",
"ExitCode": 0,
"FinishedAt": "0001-01-01T00:00:00Z",
"OOMKilled": false,
"Paused": false,
"Pid": 12428,
"Restarting": false,
"Running": true,
"StartedAt": "2015-06-08T20:47:33.459151571Z"
},
"Volumes": {
"/.swarm": "/var/lib/docker/vfs/dir/2a26044da34567f0f3601751bd1c92ceb30f2c9e6fae25043177feec85dabdc5",
"/etc/docker-ssl": "/etc/docker-ssl"
},
"VolumesRW": {
"/.swarm": true,
"/etc/docker-ssl": true
}
}
]
```

TCP resets:
```
$ sudo tcpdump -s 0 -A host dckrm01.sv2.example.com and port 2375
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 65535 bytes
13:54:40.424673 IP dckrm01.sv2.example.com.37178 > dckrs01.sv2.example.com.2375: Flags [S], seq 3761835818, win 29200, options [mss 1460,sackOK,TS val 64087891 ecr 0,nop,wscale 7], length 0
E..<.y@.>.V...S{..Q..: G.9.*......r.cd.........
...S........
13:54:40.424749 IP dckrs01.sv2.example.com.2375 > dckrm01.sv2.example.com.37178: Flags [R.], seq 0, ack 3761835819, win 0, length 0
E..(..@.?.....Q...S{ G.:.....9.+P...(i..
13:55:03.513278 IP dckrm01.sv2.example.com.37181 > dckrs01.sv2.example.com.2375: Flags [S], seq 1476755968, win 29200, options [mss 1460,sackOK,TS val 64093663 ecr 0,nop,wscale 7], length 0
E..<..@.>..[..S{..Q..= GX.........r.b3.........
............
13:55:03.513326 IP dckrs01.sv2.example.com.2375 > dckrm01.sv2.example.com.37181: Flags [R.], seq 0, ack 1476755969, win 0, length 0
E..(..@.?.....Q...S{ G.=....X...P...=...
^C
4 packets captured
7 packets received by filter
0 packets dropped by kernel 
```

**Comment by aluzzardi** [comment]
ping @abronan 

I see you are using consul. Would you mind trying this with the `swarm:0.3.0-rc2` image? We rewrote the discovery service between 0.2 and 0.3

**Comment by aluzzardi** [comment]
Also, I see in the logs:
```
time="2015-06-08T20:50:03Z" level=error msg="Get https://172.19.81.223:2375/v1.15/info: dial tcp 172.19.81.223:2375: connection refused"
```

and in your `swarm join`:
```
"--addr=172.19.81.223:2375"
```

Can the machine where swarm manage is running reach docker at that address?

**Comment by maguec** [comment]
If I run up netcat and hit it from that node, it is not firewalled

**Comment by aluzzardi** [comment]
What happens if you `docker -H 172.19.81.223:2375 info` from the machine where swarm manager is running?

**Comment by maguec** [comment]
it resets, but if on the local machine I run it
 FATA[0000] Get http://127.0.0.1:2375/v1.18/info: read tcp 127.0.0.1:2375: connection reset by peer. Are you trying to connect to a TLS-enabled daemon without TLS?

**Comment by aluzzardi** [comment]
Did you make docker listen on port `2375`?

**Comment by aluzzardi** [comment]
@moxiegirl I think this is yet another case of the getting manually started problem.

It's not obvious that swarm talks to **docker**, not to `swarm join`. It's not obvious that you must open up docker for external communication and it's not obvious that `2375` in the docs is just an example.

@maguec I don't think it's a bug, but just a docs problem. As I mentioned above, swarm **must** be able to speak to docker.

This is very helpful, if you could dedicate a few minutes: Would you mind briefly explaining what you understood from the docs/resources? We are working on improving the getting started process

**Comment by moxiegirl** [comment]
@aluzzardi I see. Just those three points. I can't promise but I will try this weekend and see if I can update the manual install with the feedback we have so far.

@maguec It would be a huge help if you could identify any areas of the doc you found awkward. Thank you for bringing this problem to our attention.

**Comment by maguec** [comment]
That is the issue.  Thanks for helping to get this running
What is confusing is that the docks say on each node do something, but the swarm should specify the "swarm manager container" specifically.  I know it makes it wordier..

**Comment by aluzzardi** [comment]
Keeping this open until we address the docs issue

**Comment by aluzzardi** [comment]
@moxiegirl: No rush - the problem existed since day 1. It's okay if we fix it even after 0.3 is out (as in, this is not a new problem introduced in 0.3)

**Comment by amitshukla** [comment]
The docs have been fixed for a while.
